python/tskit/text_formats.py

```diff
@@ -0,0 +1,85 @@
+# MIT License
+#
+# Copyright (c) 2021 Tskit Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Module responsible for working with text format data.
+"""
+import numpy as np
+
+import tskit
+
+
+def parse_fam(fam_path: str, tc: tskit.TableCollection):
+    """
+    Parse PLINK .fam file and convert to tskit IndividualTable.
+
+    Assumes fam file contains five columns: FID, IID, PAT, MAT, SEX
+
+    :param str fam_path: Path to PLINK .fam file
+    :param tskit.TableCollection tc: TableCollection with IndividualTable to
+        which the individuals will be added
+    """
+    individuals = np.loadtxt(
+        fam_path,
+        dtype=str,
+        ndmin=2,  # read file as 2-D table
+        usecols=(0, 1, 2, 3, 4),  # only keep FID, IID, PAT, MAT, SEX columns
+    )  # requires same number of columns in each row, i.e. not ragged
+
+    if len(individuals) == 0:
+        return tc
+
+    id_map = {}  # dict for translating PLINK ID to tskit IndividualTable ID
+    for tskit_id, (plink_fid, plink_iid, _pat, _mat, _sex) in enumerate(individuals):
+        # need this to guarantee uniqueness of individual ID
+        plink_id = (
+            str(plink_fid) + " " + str(plink_iid)
+        )  # include space between strings to ensure uniqueness
+        if plink_id in id_map:
+            raise ValueError("Duplicate PLINK ID")
+        id_map[plink_id] = tskit_id
+    id_map["0"] = "-1"
```

Comment:
Shouldn't I be using `setdefault` later when I'm trying to key into `id_map` with the parent IDs? For instance:
```
id_map.setdefault(pat, "-1")
```
would return the corresponding value for the key `pat`, otherwise it returns "-1"

Reply:
just a note that using setdefault will not raise a KeyError when trying to map a nonzero parent ID of a parent not present in the fam file.

Reply:
Yes, I think the right option here is to raise a KeyError if there are unknown IDs in the file. We can add options to do something different later if we need to , but the default should be to fail noisily I think.

Reply:
Sorry, my bad on the setdefault thing, I misunderstood what it did. Really the correct thing to do is just map "0" to "-1" and raise a KeyError if anything else is specified.